google/generativeai/types/retriever_types.py

```diff
@@ -0,0 +1,1319 @@
+            name: The name of the `Corpus` containing `Document`s.
+            page_size: The maximum number of `Document`s to return (per page). The service may return fewer `Document`s.
+            page_token: A page token, received from a previous `ListDocuments` call.
+
+        Return:
+            Paginated list of `Document`s.
+        """
+        if client is None:
+            client = get_default_retriever_client()
+
+        request = glm.ListDocumentsRequest(
+            parent=self.name, page_size=page_size, page_token=page_token
+        )
+        response = client.list_documents(request)
+        return response
+
+    @string_utils.set_doc(list_documents.__doc__)
+    async def list_documents_async(
+        self,
+        page_size: Optional[int] = None,
+        page_token: Optional[str] = None,
+        client: glm.RetrieverServiceAsyncClient | None = None,
+    ) -> list[Document]:
+        if client is None:
+            client = get_default_retriever_async_client()
+
+        request = glm.ListDocumentsRequest(
+            parent=self.name, page_size=page_size, page_token=page_token
+        )
+        response = await client.list_documents(request)
+        return response
+
+    def to_dict(self) -> dict[str, Any]:
+        result = {"name": self.name, "display_name": self.display_name}
+        return result
+
+
+@string_utils.prettyprint
+@dataclasses.dataclass(init=False)
+class Document(abc.ABC):
+    """
+    A `Document` is a collection of `Chunk`s.
+    """
+
+    def __init__(self, name: str, display_name: str, custom_metadata: list[CustomMetadata]):
+        self.name = name
+        self.display_name = display_name
+        self.custom_metadata = custom_metadata
+
+    def create_chunk(
+        self,
+        name: str,
+        data: ChunkData,
+        custom_metadata: Optional[list[CustomMetadata]] = None,
+        client: glm.RetrieverServiceClient | None = None,
+    ) -> Chunk:
+        """
+        Create a `Chunk` object which has textual data.
+
+        Args:
+            name: The `Chunk` resource name. The ID (name excluding the "corpora/*/documents/*/chunks/" prefix) can contain up to 40 characters that are lowercase alphanumeric or dashes (-).
+            data: The content for the `Chunk`, such as the text string.
+            custom_metadata: User provided custom metadata stored as key-value pairs.
+            state: States for the lifecycle of a `Chunk`.
+
+        Return:
+            `Chunk` object with specified data.
+
+        Raises:
+            ValueError when chunk name not specified correctly.
+        """
+        if client is None:
+            client = get_default_retriever_client()
+
+        if len(name) == 0:
+            raise ValueError("Chunk name must be specified.")
+
+        chunk = None
+        if re.match(_CHUNK_NAME_REGEX, name):
+            chunk = glm.Chunk(
+                name=name, data={"string_value": data}, custom_metadata=custom_metadata
```

Comment:
This doesn't look like it would work if this was actually a ChunkData.

Reply:
Just made another loop to handle if user passes either a str or ChunkData.